File: cpp/flyweight/flyweight.cpp
#include <iostream>
#include <memory>
#include <utility>
#include <map>

class Flyweight {
public:
    Flyweight(int id, int texture) : _id(id), _texture(texture) {
    }

    int get_id() const {
        return _id;
    }

    int get_texture() const {
        return _texture;
    }

private:
    int _id;
    int _texture;
};

class Car {
public:
    explicit Car(const Flyweight &flyweight) : _flyweight(flyweight) {}

    void draw() {
        std::cout << "Drawing texture " << _flyweight.get_texture() << " with id=" << _flyweight.get_id() << std::endl;
    }

private:
    const Flyweight &_flyweight;
};

class CarFactory {
public:
    Car create_car(int texture_config) {
        if (_flyweights.find(texture_config) == _flyweights.cend()) {
            _flyweights[texture_config] = std::make_shared<Flyweight>(rand(), texture_config);
        }
        return Car(*_flyweights[texture_config]);
    }

private:
    std::map<int, std::shared_ptr<Flyweight>> _flyweights;
};

int main() {
    CarFactory factory;
    Car car1 = factory.create_car(1);
    Car car2 = factory.create_car(1);
    Car car3 = factory.create_car(5);
    Car car4 = factory.create_car(1);
    car1.draw();
    car2.draw();
    car3.draw();
    car4.draw();
    return 0;
}

File: python/flyweight/flyweight.py
from random import randrange


class Flyweight:
    def __init__(self, id_: int, texture: int):
        self._id_: int = id_
        self._texture: int = texture

    def get_id(self) -> int:
        return self._id_

    def get_texture(self) -> int:
        return self._texture


class Car:
    def __init__(self, flyweight: Flyweight):
        self._flyweight = flyweight

    def draw(self):
        print(f"Drawing car with texture {self._flyweight.get_texture()} and id={self._flyweight.get_id()}")


class CarFactory:
    def __init__(self):
        self._flyweights = {}

    def create(self, texture_config: int) -> Car:
        if texture_config not in self._flyweights:
            self._flyweights[texture_config] = Flyweight(randrange(0, 1000), texture_config)
        return Car(self._flyweights[texture_config])


def main():
    factory = CarFactory()
    car1: Car = factory.create(3)
    car2: Car = factory.create(3)
    car3: Car = factory.create(1)
    car4: Car = factory.create(3)
    car1.draw()
    car2.draw()
    car3.draw()
    car4.draw()


if __name__ == '__main__':
    main()
